pretty_gpx/drawing/drawing_city.py

```diff
@@ -0,0 +1,124 @@
+import numpy as np
```

Comment:
@JulesL2 
Great for park and forest! :)

I've pushed a big refacto yesterday to allow to use VerticalLayout and DrawingFigure with other modes than the mountain elevation map. I'll soon init the city part

I've tested your code. Good news the display with LineCollection is alsmot instantaneous! however the query is quite slow (from 10s to 1min), which can be cached like elevation map using the new GpxDataCacheHandler. I briefly tried with osmnx to speed it up, but it's slower since it builds the graph of roads.  Maybe there's something smart to do by parallelizing or paginating the overpy requests :thinking:

Reply:
I have tryied multiple queries for the roads but since we need all nodes I believe it is what it is.
The thing we need to be carefull with Overpass API is that if we need nodes, we need to always add the recursion in the request otherwise, overpy will create a new request for each node missing.

I have seen the refacto, I can try to init cities this afternoon in this PR :)

> Maybe there's something smart to do by parallelizing or paginating the overpy requests 🤔

For the roads as we use multiple priority level (for different type of roads) we could launch the 4 requests in parralel instead of sequential, that will surely reduce by at least 2 the time needed.

Reply:
And if you have any ideas of other things to add on the map (I am looking for airport tracks right now)

Reply:
I've put current changes on this branch, this is Work-in-progress but it can still help you to quickly set up the cache and layout
I was playing around with the color themes
https://github.com/ThomasParistech/pretty-gpx/tree/draw-city-roads

Reply:
@ThomasParistech I just push the modifications that I have made based on your branch
I did:

-> Overpass API : fix the recursion in the request (it was done only on the last item of the request and not on the whole group

-> utils Add fonctions to support mercator projection

-> gpxtrack Add fonctions to support mercator projection and activity time

-> city_vertical_layout Add a function to get the bounds but with respect to mercator coordinates

-> city_roads Add functions to support rivers and forests using relations and fonctions to process relations members


Here are the patches if you want to continue to work on your branch but with these development: (git am path/to/patch/000*.patch apply them)


[0004-add-functions-to-get-rivers-and-forests.patch](https://github.com/user-attachments/files/17029569/0004-add-functions-to-get-rivers-and-forests.patch)
[0005-sparkles-add-functions-to-get-rivers-and-forests-2.patch](https://github.com/user-attachments/files/17029571/0005-sparkles-add-functions-to-get-rivers-and-forests-2.patch)

Reply:
@ThomasParistech 
Quick update, I have add the functions to get airports and railways
(This patch)
[0006-sparkles-add-functions-to-get-rivers-and-forests-3.patch](https://github.com/user-attachments/files/17042923/0006-sparkles-add-functions-to-get-rivers-and-forests-3.patch)

Here is the result for now (we need to zoom or look at train stations to see rails, I will look more deeply in the drawing settings later) :

I will try to implement my drawing function inside the drawing class

![output](https://github.com/user-attachments/assets/9d47f141-e177-44c4-af7b-eb17febbfcfe)

Reply:
@JulesL2 
looks great! :sparkles: 

One remark though; since the overpy request or the elevation map is retrieved on a lat/lon bbox, i'd prefer to keep working in lat/lon coordinates, and make the reasonable assumption that the earth surface is locally planar to use a fixed aspect ratio for the whole area. Always working in mercator coordinates would make a tiny difference in rendering but would force us to remap the rectangle elevation map to its distorted projection. It allows better factorization between mountain and city modes. As was done in https://github.com/ThomasParistech/pretty-gpx/compare/da421d4cae95cb711e447fcc8c583c9a20617cc8...draw-city-roads

Reply:
Problem is the earth is flat locally but as Paris is not at the equator we still have a distored map when using lat/lon (here I made the test and I let you see that it is clearly less pretty) That's why I chose the mercator, the objective was that people would more easily recognize the city. And plotting images on top of it such as monuments minatures (i'm looking forward to develop this) can be done only on "flat" mercator maps
![output](https://github.com/user-attachments/assets/58df4ad0-e53a-4992-9f1e-c6440e0e938a)

Reply:
i got this image with the lat/lon coordinates when running the script yyyyyyy2.py from my wip branch, which looks the same as yours with mercator. I suspect your image to be plotted with the wrong aspect ratio :thinking: 

![image](https://github.com/user-attachments/assets/415d1aa4-5be5-4f98-abcf-b9d909355741)

Reply:
Yeah my bad, forgot to change the aspect ratio of the ax that was set to equal with mercator.
Then yeah I will remove all these functions in a patch :)

Reply:
I've just merged the roads collection and added a wip snippet to run `https://github.com/ThomasParistech/pretty-gpx/blob/main/pretty_gpx/city/wip_city_sample.py`

I did not include the forest/river/airport/railway as they do not look mature yet

Could you clean the forest/river extraction and add their specific files like roads.py and make sure the type checker is happy (run `sh scripts/run_mypy.sh`) ? Some of the code can be put in `pretty_gpx/common/data/overpass_processing.py`

By the way, how did you manage the interior holes of shapely polygons when plotting them? Could you convert them to the PolyFillData class or add an equivalent ?

Many thanks :fire:

Reply:
Hey,
Thank you for the merge, I will re-do  forest/river/rails ect properly and will look on a class to plot properly the polygons (I am already using a class in a notebook so it should not be a problem.